File: src/sc_neurocore/accel/go/services/perfect_integrator.go
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial license available
// © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
// © Code 2020–2026 Miroslav Šotek. All rights reserved.
// ORCID: 0009-0009-3560-0851
// Contact: www.anulum.li | protoscience@anulum.li
// SC-NeuroCore — Go service for perfect_integrator

package services

import (
	"errors"
	"math"
)

// PerfectIntegratorNeuronState holds the neuron state.
type PerfectIntegratorNeuronState struct {
	V          float64
	CM         float64
	VThreshold float64
	VReset     float64
	Dt         float64
	// SourceProfile selects Naud-Gerstner's strict > threshold boundary.
	SourceProfile bool
}

// NewNaudGerstnerPerfectIntegrator creates the strict source-equation profile.
func NewNaudGerstnerPerfectIntegrator() *PerfectIntegratorNeuronState {
	state := NewPerfectIntegratorNeuron()
	state.SourceProfile = true
	return state
}

// NewPerfectIntegratorNeuron creates a new PerfectIntegratorNeuron neuron with default parameters.
func NewPerfectIntegratorNeuron() *PerfectIntegratorNeuronState {
	return &PerfectIntegratorNeuronState{
		V:          0.0,
		CM:         1.0,
		VThreshold: 1.0,
		VReset:     0.0,
		Dt:         0.1,
	}
}

// Valid reports whether the state satisfies the non-leaky integration contract.
func (s PerfectIntegratorNeuronState) Valid() bool {
	return finitePerfectIntegrator(s.V) &&
		finitePerfectIntegrator(s.CM) && s.CM > 0.0 &&
		finitePerfectIntegrator(s.VThreshold) &&
		finitePerfectIntegrator(s.VReset) && s.VThreshold > s.VReset &&
		((s.SourceProfile && s.V <= s.VThreshold) ||
			(!s.SourceProfile && s.V < s.VThreshold)) &&
		finitePerfectIntegrator(s.Dt) && s.Dt > 0.0
}

// Step advances the neuron by one timestep. Invalid inputs do not mutate state.
func (s *PerfectIntegratorNeuronState) Step(iExt float64) (int, error) {
	if !finitePerfectIntegrator(iExt) || !s.Valid() {
		return 0, ErrPerfectIntegratorInvalidState
	}

	voltageIncrement := iExt * s.Dt / s.CM
	nextV := s.V + voltageIncrement
	if !finitePerfectIntegrator(voltageIncrement) || !finitePerfectIntegrator(nextV) {
		return 0, ErrPerfectIntegratorNonFiniteUpdate
	}

	crossed := nextV >= s.VThreshold
	if s.SourceProfile {
		crossed = nextV > s.VThreshold
	}
	if crossed {
		s.V = s.VReset
		return 1, nil
	}
	s.V = nextV
	return 0, nil
}

// Reset restores dynamic state without changing parameters.
func (s *PerfectIntegratorNeuronState) Reset() {
	s.V = s.VReset
}

// SimulatePerfectIntegratorNeuron runs the neuron for n steps.
func SimulatePerfectIntegratorNeuron(nSteps int, iExt float64) ([]float64, int) {
	trace, spikes, err := SimulatePerfectIntegratorTrace(
		*NewPerfectIntegratorNeuron(), nSteps, iExt,
	)
	if err != nil {
		panic(err)
	}
	return trace, spikes
}

// SimulatePerfectIntegratorTrace executes a complete state/parameter contract.
func SimulatePerfectIntegratorTrace(
	initial PerfectIntegratorNeuronState,
	nSteps int,
	iExt float64,
) ([]float64, int, error) {
	if nSteps < 0 || !finitePerfectIntegrator(iExt) || !initial.Valid() {
		return nil, 0, ErrPerfectIntegratorInvalidState
	}
	state := initial
	trace := make([]float64, nSteps)
	spikes := 0
	for index := 0; index < nSteps; index++ {
		spike, err := state.Step(iExt)
		if err != nil {
			return nil, 0, err
		}
		spikes += spike
		trace[index] = state.V
	}
	return trace, spikes, nil
}

// SimulatePerfectIntegratorComplete returns an aligned, failure-atomic packet.
func SimulatePerfectIntegratorComplete(
	initial PerfectIntegratorNeuronState,
	nSteps int,
	iExt float64,
) ([]float64, []uint8, float64, error) {
	if nSteps < 0 || !finitePerfectIntegrator(iExt) || !initial.Valid() {
		return nil, nil, initial.V, ErrPerfectIntegratorInvalidState
	}
	state := initial
	trace := make([]float64, nSteps)
	events := make([]uint8, nSteps)
	for index := 0; index < nSteps; index++ {
		event, err := state.Step(iExt)
		if err != nil {
			return nil, nil, initial.V, err
		}
		trace[index] = state.V
		events[index] = uint8(event)
	}
	return trace, events, state.V, nil
}

func finitePerfectIntegrator(value float64) bool {
	return !math.IsNaN(value) && !math.IsInf(value, 0)
}

var (
	ErrPerfectIntegratorInvalidState    = errors.New("perfect-integrator state/current must be finite and physically ordered")
	ErrPerfectIntegratorNonFiniteUpdate = errors.New("perfect-integrator voltage increment became non-finite")
)

File: src/sc_neurocore/accel/julia/neurons/perfect_integrator.jl
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SC-NeuroCore — Julia for perfect_integrator

module PerfectIntegratorAccel

export step!, simulate, simulate_trace, PerfectIntegratorNeuronState, valid, reset!

mutable struct PerfectIntegratorNeuronState
    v::Float64
    c_m::Float64
    v_threshold::Float64
    v_reset::Float64
    dt::Float64
end

function PerfectIntegratorNeuronState()
    PerfectIntegratorNeuronState(0.0, 1.0, 1.0, 0.0, 0.1)
end

function valid(s::PerfectIntegratorNeuronState)::Bool
    return isfinite(s.v) &&
        isfinite(s.c_m) && s.c_m > 0.0 &&
        isfinite(s.v_threshold) &&
        isfinite(s.v_reset) && s.v_threshold > s.v_reset &&
        s.v < s.v_threshold &&
        isfinite(s.dt) && s.dt > 0.0
end

function step!(s::PerfectIntegratorNeuronState, I_ext::Float64=0.0; dt::Float64=s.dt)::Int
    if !isfinite(I_ext)
        throw(DomainError(I_ext, "PerfectIntegrator input current must be finite"))
    end
    if !isfinite(dt) || dt <= 0.0
        throw(DomainError(dt, "PerfectIntegrator dt must be finite and positive"))
    end
    if !valid(s)
        throw(DomainError((s.v, s.c_m, s.v_threshold, s.v_reset, s.dt), "PerfectIntegrator state must be finite and physically ordered"))
    end

    voltage_increment = I_ext / s.c_m * dt
    next_v = s.v + voltage_increment
    if !isfinite(voltage_increment) || !isfinite(next_v)
        throw(DomainError((voltage_increment, next_v), "PerfectIntegrator voltage increment must remain finite"))
    end

    s.v = next_v
    if s.v >= s.v_threshold
        s.v = s.v_reset
        return 1
    end
    return 0
end

function reset!(s::PerfectIntegratorNeuronState)::Nothing
    s.v = s.v_reset
    return nothing
end

function simulate(n_steps::Int=1000; I_ext::Float64=10.0, dt::Float64=0.1)
    result = simulate_trace(0.0, 1.0, 1.0, 0.0, dt, n_steps, I_ext)
    return result.trace, result.spikes
end

function simulate_trace(
    v::Float64,
    c_m::Float64,
    v_threshold::Float64,
    v_reset::Float64,
    dt::Float64,
    n_steps::Int,
    I_ext::Float64,
)
    if n_steps < 0
        throw(ArgumentError("PerfectIntegrator n_steps must be non-negative"))
    end
    if !isfinite(I_ext)
        throw(DomainError(I_ext, "PerfectIntegrator input current must be finite"))
    end
    s = PerfectIntegratorNeuronState(v, c_m, v_threshold, v_reset, dt)
    if !valid(s)
        throw(DomainError(v, "PerfectIntegrator state must be finite and physically ordered"))
    end
    trace = zeros(n_steps)
    spikes = 0
    for t in 1:n_steps
        result = step!(s, I_ext)
        trace[t] = s.v
        if result > 0
            spikes += 1
        end
    end
    return (trace=trace, spikes=spikes, vf=s.v)
end

end # module PerfectIntegratorAccel
